File: java/src/main/java/benchmark/serializers/Versions.java
package benchmark.serializers;

import java.io.InputStream;
import java.util.Properties;

/**
 * Best-effort library version. Prefer Maven-filtered versions.properties; fall back to package
 * Implementation-Version (often missing in shaded jars).
 */
public final class Versions {
  private static final Properties PROPS = load();

  private Versions() {}

  private static Properties load() {
    Properties p = new Properties();
    try (InputStream in =
        Versions.class.getClassLoader().getResourceAsStream("benchmark-versions.properties")) {
      if (in != null) p.load(in);
    } catch (Exception ignored) {
      // optional resource
    }
    return p;
  }

  public static String of(Class<?> cls) {
    String key = cls.getName();
    String v = PROPS.getProperty(key);
    if (v != null && !v.isBlank()) return v;
    // try simple name keys
    v = PROPS.getProperty(cls.getSimpleName());
    if (v != null && !v.isBlank()) return v;
    Package p = cls.getPackage();
    if (p != null) {
      String iv = p.getImplementationVersion();
      if (iv != null && !iv.isBlank()) return iv;
    }
    return PROPS.getProperty("default", "unknown");
  }
}

File: kotlin/src/main/kotlin/benchmark/serializers/Versions.kt
package benchmark.serializers

import java.util.Properties

/**
 * Best-effort library version. Prefer filtered versions.properties; fall back to package
 * Implementation-Version (often missing in shaded jars).
 */
object Versions {
    private val props: Properties = load()

    private fun load(): Properties {
        val p = Properties()
        try {
            Versions::class.java.classLoader.getResourceAsStream("benchmark-versions.properties")?.use {
                p.load(it)
            }
        } catch (_: Exception) {
            // optional resource
        }
        return p
    }

    fun of(cls: Class<*>): String {
        val key = cls.name
        var v = props.getProperty(key)
        if (!v.isNullOrBlank()) return v
        v = props.getProperty(cls.simpleName)
        if (!v.isNullOrBlank()) return v
        val iv = cls.`package`?.implementationVersion
        if (!iv.isNullOrBlank()) return iv
        return props.getProperty("default", "unknown")
    }

    fun of(key: String, fallbackClass: Class<*>? = null): String {
        val v = props.getProperty(key)
        if (!v.isNullOrBlank()) return v
        return if (fallbackClass != null) of(fallbackClass) else props.getProperty("default", "unknown")
    }
}
